File: ios/RNVideoExample/ShakaPlayerModule.m
#import <Foundation/Foundation.h>
#import <React/RCTBridgeModule.h>
#import <React/RCTViewManager.h>

@interface RCT_EXTERN_MODULE(ShakaPlayerModule, NSObject)

RCT_EXTERN_METHOD(createPlayer:(nonnull NSNumber *)viewId)
RCT_EXTERN_METHOD(destroyPlayer:(nonnull NSNumber *)viewId)
RCT_EXTERN_METHOD(loadVideo:(nonnull NSNumber *)viewId url:(nonnull NSString *)url)
RCT_EXTERN_METHOD(play:(nonnull NSNumber *)viewId)
RCT_EXTERN_METHOD(pause:(nonnull NSNumber *)viewId)
RCT_EXTERN_METHOD(seekTo:(nonnull NSNumber *)viewId position:(nonnull NSNumber *)position)
RCT_EXTERN_METHOD(setVolume:(nonnull NSNumber *)viewId volume:(nonnull NSNumber *)volume)
RCT_EXTERN_METHOD(getCurrentPosition:(nonnull NSNumber *)viewId
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(getDuration:(nonnull NSNumber *)viewId
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(getIsPlaying:(nonnull NSNumber *)viewId
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

+ (BOOL)requiresMainQueueSetup
{
  return NO;
}

@end

@interface RCT_EXTERN_MODULE(ShakaPlayerViewManager, RCTViewManager)

RCT_EXPORT_VIEW_PROPERTY(onReady, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onPlay, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onPause, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onBuffer, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onError, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onProgress, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onEnd, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(showNativeControls, BOOL)

@end

File: ios/RNVideoExample/ShakaPlayerModule.swift
import Foundation
import AVFoundation
import React

@objc(ShakaPlayerModule)
class ShakaPlayerModule: NSObject {
    private var players: [Int: AVPlayer] = [:]
    private var observers: [Int: Any] = [:]
    private var views: [Int: ShakaPlayerView] = [:]
    
    func registerPlayer(_ player: AVPlayer, forViewId viewId: Int) {
        DispatchQueue.main.async {
            self.players[viewId] = player
        }
    }
    
    func registerView(_ view: ShakaPlayerView, forViewId viewId: Int) {
        DispatchQueue.main.async {
            self.views[viewId] = view
        }
    }
    
    @objc
    func createPlayer(_ viewId: NSNumber) {
        // Player is created by the view, this just ensures compatibility
    }
    
    @objc
    func destroyPlayer(_ viewId: NSNumber) {
        let id = viewId.intValue
        DispatchQueue.main.async {
            if let observer = self.observers[id] {
                self.players[id]?.removeTimeObserver(observer)
                self.observers.removeValue(forKey: id)
            }
            self.players.removeValue(forKey: id)
            self.views.removeValue(forKey: id)
        }
    }
    
    @objc
    func loadVideo(_ viewId: NSNumber, url: NSString) {
        let id = viewId.intValue
        guard let player = players[id],
              let videoURL = URL(string: url as String) else {
            return
        }
        
        DispatchQueue.main.async {
            let playerItem = AVPlayerItem(url: videoURL)
            player.replaceCurrentItem(with: playerItem)
            
            // Notify view to update item observers
            self.views[id]?.updatePlayerItem()
        }
    }
    
    @objc
    func play(_ viewId: NSNumber) {
        let id = viewId.intValue
        guard let player = players[id] else { return }
        
        DispatchQueue.main.async {
            player.play()
        }
    }
    
    @objc
    func pause(_ viewId: NSNumber) {
        let id = viewId.intValue
        guard let player = players[id] else { return }
        
        DispatchQueue.main.async {
            player.pause()
        }
    }
    
    @objc
    func seekTo(_ viewId: NSNumber, position: NSNumber) {
        let id = viewId.intValue
        guard let player = players[id] else { return }
        
        let time = CMTime(seconds: position.doubleValue, preferredTimescale: 1000)
        DispatchQueue.main.async {
            player.seek(to: time)
        }
    }
    
    @objc
    func setVolume(_ viewId: NSNumber, volume: NSNumber) {
        let id = viewId.intValue
        guard let player = players[id] else { return }
        
        DispatchQueue.main.async {
            player.volume = volume.floatValue
        }
    }
    
    @objc
    func getCurrentPosition(_ viewId: NSNumber,
                           resolve: @escaping RCTPromiseResolveBlock,
                           reject: @escaping RCTPromiseRejectBlock) {
        let id = viewId.intValue
        guard let player = players[id] else {
            reject("ERROR", "Player not found", nil)
            return
        }
        
        DispatchQueue.main.async {
            let currentTime = player.currentTime().seconds
            resolve(currentTime)
        }
    }
    
    @objc
    func getDuration(_ viewId: NSNumber,
                    resolve: @escaping RCTPromiseResolveBlock,
                    reject: @escaping RCTPromiseRejectBlock) {
        let id = viewId.intValue
        guard let player = players[id],
              let duration = player.currentItem?.duration else {
            reject("ERROR", "Player or duration not found", nil)
            return
        }
        
        DispatchQueue.main.async {
            let durationSeconds = duration.seconds
            if durationSeconds.isNaN {
                resolve(0)
            } else {
                resolve(durationSeconds)
            }
        }
    }
    
    @objc
    func getIsPlaying(_ viewId: NSNumber,
                     resolve: @escaping RCTPromiseResolveBlock,
                     reject: @escaping RCTPromiseRejectBlock) {
        let id = viewId.intValue
        guard let player = players[id] else {
            reject("ERROR", "Player not found", nil)
            return
        }
        
        DispatchQueue.main.async {
            resolve(player.rate > 0)
        }
    }
    
    @objc
    static func requiresMainQueueSetup() -> Bool {
        return false
    }
}
